functable.c

```diff
@@ -157,190 +160,192 @@ static void init_functable(void) {
+        ft->update_hash = &update_hash_acle;
     }
 #endif
 
 
     // Power - VMX
 #ifdef PPC_VMX
     if (cf.power.has_altivec) {
-        ft.adler32 = &adler32_vmx;
-        ft.slide_hash = &slide_hash_vmx;
+        ft->adler32 = &adler32_vmx;
+        ft->slide_hash = &slide_hash_vmx;
     }
 #endif
     // Power8 - VSX
 #ifdef POWER8_VSX
     if (cf.power.has_arch_2_07) {
-        ft.adler32 = &adler32_power8;
-        ft.chunkmemset_safe = &chunkmemset_safe_power8;
-        ft.chunksize = &chunksize_power8;
-        ft.inflate_fast = &inflate_fast_power8;
-        ft.slide_hash = &slide_hash_power8;
+        ft->adler32 = &adler32_power8;
+        ft->chunkmemset_safe = &chunkmemset_safe_power8;
+        ft->chunksize = &chunksize_power8;
+        ft->inflate_fast = &inflate_fast_power8;
+        ft->slide_hash = &slide_hash_power8;
     }
 #endif
 #ifdef POWER8_VSX_CRC32
     if (cf.power.has_arch_2_07)
-        ft.crc32 = &crc32_power8;
+        ft->crc32 = &crc32_power8;
 #endif
     // Power9
 #ifdef POWER9
     if (cf.power.has_arch_3_00) {
-        ft.compare256 = &compare256_power9;
-        ft.longest_match = &longest_match_power9;
-        ft.longest_match_slow = &longest_match_slow_power9;
+        ft->compare256 = &compare256_power9;
+        ft->longest_match = &longest_match_power9;
+        ft->longest_match_slow = &longest_match_slow_power9;
     }
 #endif
 
 
     // RISCV - RVV
 #ifdef RISCV_RVV
     if (cf.riscv.has_rvv) {
-        ft.adler32 = &adler32_rvv;
-        ft.adler32_fold_copy = &adler32_fold_copy_rvv;
-        ft.chunkmemset_safe = &chunkmemset_safe_rvv;
-        ft.chunksize = &chunksize_rvv;
-        ft.compare256 = &compare256_rvv;
-        ft.inflate_fast = &inflate_fast_rvv;
-        ft.longest_match = &longest_match_rvv;
-        ft.longest_match_slow = &longest_match_slow_rvv;
-        ft.slide_hash = &slide_hash_rvv;
+        ft->adler32 = &adler32_rvv;
+        ft->adler32_fold_copy = &adler32_fold_copy_rvv;
+        ft->chunkmemset_safe = &chunkmemset_safe_rvv;
+        ft->chunksize = &chunksize_rvv;
+        ft->compare256 = &compare256_rvv;
+        ft->inflate_fast = &inflate_fast_rvv;
+        ft->longest_match = &longest_match_rvv;
+        ft->longest_match_slow = &longest_match_slow_rvv;
+        ft->slide_hash = &slide_hash_rvv;
     }
 #endif
 
 
     // S390
 #ifdef S390_CRC32_VX
     if (cf.s390.has_vx)
-        ft.crc32 = crc32_s390_vx;
+        ft->crc32 = crc32_s390_vx;
 #endif
+}
 
+/* Platform has pointer size atomic store */
+#if defined(__GNUC__) || defined(__clang__)
+#  define FUNCTABLE_ASSIGN(VAR, FUNC_NAME) \
+    __atomic_store(&(functable.FUNC_NAME), &(VAR->FUNC_NAME), __ATOMIC_SEQ_CST)
+#elif defined(_MSC_VER)
+#  define FUNCTABLE_ASSIGN(VAR, FUNC_NAME) \
+    _InterlockedExchangePointer((void * volatile *)&(functable.FUNC_NAME), (void *)(VAR->FUNC_NAME))
+#else
+    (void * volatile)(functable.FUNC_NAME) = (void *)(VAR->FUNC_NAME)
```

Comment:
Do you have a reference that confirms that this is safe?

Reply:
Sorry, I don't have a public link to the discussion :(